api/v1alpha1/hostedcluster_types.go

```diff
@@ -846,10 +837,134 @@ type AWSResourceTag struct {
 	Value string `json:"value"`
 }
 
-type AWSRoleCredentials struct {
-	ARN       string `json:"arn"`
-	Namespace string `json:"namespace"`
-	Name      string `json:"name"`
+// AWSRoles contains references to various AWS IAM roles required to enable
+// integrations such as OIDC.
+type AWSRoles struct {
+	// IngressOIDC is an ARN value referencing a role used for ingress OIDC
+	// integration.
+	//
+	// The following is an example of a valid policy document:
+	//
+	// {
+	//	"Version": "2012-10-17",
+	//	"Statement": [
+	//		{
+	//			"Effect": "Allow",
+	//			"Action": [
+	//				"elasticloadbalancing:DescribeLoadBalancers",
+	//				"tag:GetResources",
+	//				"route53:ListHostedZones"
+	//			],
+	//			"Resource": "*"
+	//		},
+	//		{
+	//			"Effect": "Allow",
+	//			"Action": [
+	//				"route53:ChangeResourceRecordSets"
+	//			],
+	//			"Resource": [
+	//				"arn:aws:route53:::PUBLIC_ZONE_ID",
+	//				"arn:aws:route53:::PRIVATE_ZONE_ID"
+	//			]
+	//		}
+	//	]
+	// }
+	IngressOIDC string `json:"ingressOidc"`
```

Comment:
In general I really like to have this properly typed. But what will we do if the set of required roles changes across OCP versions? So far, we went with "always create everything any known OCP version needs", do we want to continue like that?

Reply:
Does the array somehow account for that? It still requires the user know magic names to correspond an ARN with its purpose, and the list contents may need to change across versions in the scenario you describe.

Is the difference thus aesthetics/ergonomics? e.g. if a new field becomes required, either the user has to start adding a new magic name to the list (in the array), or we have to add a new optional field (in the struct)?

Is there a certain example you had in mind that we can use to compare how either approach holds up?

Reply:
@alvaroaleman good point. Once thought is if the assumptions ever change we could let all these fields become optional and have backend validation per ocp version.

Reply:
Yeah, makes sense, just having all fields we might need for any version and then validating seems fine, still a net improvement over what we have today